cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,441 @@
+                    // So we need to abort the send task, and notify the process input
+                    // task
+                    co_await set_ch_idx_invalid.set_channel_idx_invalid();
+                    co_return;
+                }
+            }
+            logger.trace(
+                "sent ", self, " [", ch_next_idx[self], ", ", curr_recv_msg_sz, ")"
+            );
+
+            // now next_idx can be updated to end_idx, and if !input_done, we need to
+            // request the recv task for more data
+            auto lock = co_await mtx.scoped_lock();
+            ch_next_idx[self] = curr_recv_msg_sz;
+            if (ch_next_idx[self] == recv_messages.size()) {
+                if (input_done) {
+                    // no more messages will be received, and all messages have been sent
+                    break;
+                } else {
+                    // request more data from the input channel
+                    lock.unlock();
+                    co_await request_data.notify_one();
+                }
+            }
+        }
+
+        co_await ch_out->drain(ctx.executor());
+        logger.trace("Send task ", self, " completed");
+    }
+
+    /**
+     * @brief RAII helper class to set input_done to true and notify all send tasks to
+     * wind down when the unbounded fanout state goes out of scope.
+     */
+    struct SetInputDoneAtExit {
+        UnboundedFanout* fanout;
+
+        ~SetInputDoneAtExit() {
+            coro::sync_wait(set_input_done());
+        }
+
+        // forcibly set input_done to true and notify all send tasks to wind down
+        Node set_input_done() {
+            {
+                auto lock = co_await fanout->mtx.scoped_lock();
+                fanout->input_done = true;
+            }
+            co_await fanout->data_ready.notify_all();
+        }
+    };
+
+    /**
+     * @brief Wait for a data request from the send tasks.
+     *
+     * @return The index of the last completed message and the index of the latest
+     * processed message. If both are InvalidIdx, it means that all send tasks are in an
+     * invalid state.
+     */
+    auto wait_for_data_request() -> coro::task<std::pair<size_t, size_t>> {
+        size_t last_completed_idx = InvalidIdx;
+        size_t latest_processed_idx = InvalidIdx;
+
+        auto lock = co_await mtx.scoped_lock();
+        co_await request_data.wait(lock, [&] {
+            auto filtered_view = std::ranges::filter_view(ch_next_idx, [](size_t idx) {
+                return idx != InvalidIdx;
+            });
+
+            auto it = std::ranges::begin(filtered_view);  // first valid idx
+            auto end = std::ranges::end(filtered_view);  // end idx
+
+            if (it == end) {
+                // no valid indices, so all send tasks are in an invalid state
+                return true;
+            }
+
+            auto [min_it, max_it] = std::minmax_element(it, end);
+            last_completed_idx = *min_it;
```

Comment:
"last" seems like the wrong verb here. I think this is the "lowest" completed index?